lib/jellyfish/room.ex

```diff
@@ -295,6 +299,79 @@ defmodule Jellyfish.Room do
     {:noreply, state}
   end
 
+  @impl true
+  def handle_info({:room_created, node}, %__MODULE__{id: id} = state) do
+    if node != Node.self() do
+      Phoenix.PubSub.direct_broadcast(node, Jellyfish.PubSub, id, {:room_exists, Node.self()})
+      Logger.info("#{node} created multipart room (room_id=#{id})")
+    end
+
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info({:room_exists, node}, state) do
+    if node != Node.self() do
+      Logger.info("Room #{state.id} exists on node #{node}")
+
+      endpoint_id = UUID.uuid4()
+
+      Engine.add_endpoint(
+        state.engine_pid,
+        %Remote{rtc_engine: state.engine_pid, owner: self()},
+        id: endpoint_id
+      )
+
+      {:noreply, put_in(state, [:nodes, endpoint_id], node)}
+    else
+      {:noreply, state}
+    end
+  end
+
+  @impl true
+  def handle_info(%Message.EndpointAdded{}, state), do: {:noreply, state}
+
+  @impl true
+  def handle_info(
+        %Message.EndpointMessage{
+          endpoint_id: endpoint_id,
+          endpoint_type: Remote,
+          message: {:link_proposal, %Remote.LinkProposal{} = link_proposal}
+        },
+        %__MODULE__{id: room_id} = state
+      ) do
+    case get_in(state, [:nodes, endpoint_id]) do
+      nil ->
+        Logger.warning("Wrong Remote Endpoint")
+
+      node ->
+        Phoenix.PubSub.direct_broadcast(
+          node,
+          Jellyfish.PubSub,
+          room_id,
+          {:create_remote_endpoint, link_proposal, Node.self()}
+        )
+    end
+
+    {:noreply, state}
+  end
```

Comment:
The whole logic could be simplified. When a user provides the id of the room, we know that he wants to create a distributed. In this situation, we add a remote endpoint and broadcast message to other rooms from this endpoint that we start distributed jellyroom. Other rooms receive a message about starting a distributed room, and they add remoteEndpoint if it doesn't already exist. The biggest drawback of the current implementation is that we have to maintain information about nodes and in a situation of node crash we have to make it consistent.

Reply:
That logic also causes troubles:

1. We actually don't know if a user is creating a distributed room. API allows to pass ID but it's not checking if the room with this ID exists on other nodes (it's OK, it shouldn't be API's responsibility). If we decide to check it somewhere it'll be the same logic, but moved. 

2. In this implementation I decided to create Remote Endpoints as pairs. Creating RE without info that a room exists somewhere else makes no sense because it'll cause us to implement new logic when the room is distributed to more than two nodes.

IMO it's better to create an RE in response to info that a room exists, even if we need to handle node crash. I'm not sure if need to keep information about nodes. I left it in room's state, but the most actual information about linked nodes is in the RE state. Also RE has a logic to update partner node address by sending a handshake message at any time. Maybe we should remove nodes info from room state as it'll probably be useless after linking - all further messages will be send via RE

Reply:
1. When the client wants to create a second engine on a different node, he always has to pass ID while creating a room on an other node. As for passing ID, while creating the first engine, I don't see much problem because the only cost we will get is to spawn an additional endpoint that will dump every received message, so this is not a problem. Also, the whole beauty of this solution is that we don't have to check if the room exists on the other node. We only inform other nodes that we are ready for this room to work distributed.
2. I don't understand this point, why we would need different logic.

Yeah I agree about that jellyfish should have minimal information about nodes and if anywhere it should be moved to RE.